README
# NAME

libpenmon - Implementation of Penman-Monteith equation in C++

# WARNING

this library is work-in-progress. Proceed with caution!

# DESCRIPTION

Implements a weather station class (Station) that calculates 
ETo (reference evapotranspiration) given climate data.

# SYNOPSIS

    #include "libpenmon.h"
    #include <iostream>
    
    int main(void) {
        Station station(41.42, 109);
        StationDay day(238, station);
        day.temp_min=19.5;
        day.temp_max=29.5;
        std::cout << "ETo is " << day.eto() << "mm\n";
        return 0;
    }

# AUTHOR

Sherzod B. Ruzmetov <sherzodr@gmail.com>



ChangeLog
[1.0] - 2020-10-24

First release with autotools


libpenmon.h
/*
 * libpenmon.h
 *
 *  Created on: Oct 9, 2020
 *      Author: sherzodr
 */

#ifndef LIBPENMON_H_
#define LIBPENMON_H_


const bool CHECK_RADIATION_RANGE = true;
const bool CHECK_SUNSHINE_HOURS_RANGE=true;

#include <math.h>

// forward declarations
class Station;
class StationDay;
// end forward declarations

class Crop {
public:
	int resistance_a;
	float albedo;
	float height;
	Crop(void);
};

class Climate {
public:
	bool interior_location;
	bool coastal_location;
	bool island_location;

	bool arid_climate;
	bool humid_climate;

	short dew_point_difference;
	float average_wind_speed;
	float k_rs = 0.16;

	Climate(void);
	explicit operator bool() const { return true; }

	void light_winds();
	void moderate_winds();
	void strong_winds();

	void arid();

	void humid();

	void interior();
	void coastal();
	void island();
};

class Station {
public:

	float latitude, latitude_radians;
	int altitude;
	short anemometer_height = 2;

	Crop ref_crop;
	Climate climate;

	explicit operator bool() const { return true; }
	Station(float lat, int alt);
	Station(float lat, int alt, short anem_height);
	StationDay get_day(short d);
};

class StationDay {

public:

	short day_number;
	float temp_min { }, temp_max { }, temp_mean { }, humidity_mean { },
			humidity_min { }, humidity_max { };
	float wind_speed { }, radiation_s { }, radiation_a { }, temp_dew { },
			temp_dry { }, temp_wet { };
	float vapour_pressure { }, sunshine_hours {-1 };
	const float stephen_boltzmann_constant = 4.903 * pow(10, -9);

	Station* station;

	StationDay(short d, Station* st);

	float wind_speed_2m(void);
	float dew_point(void);
	float atmospheric_pressure(void);
	float atmospheric_pressure_in_mm(void);
	float latent_heat_of_vapourization(void);
	double specific_heat(void);
	float psychrometric_constant(void);
	float Tmean(void);
	float saturation_vapour_pressure(float);
	float mean_saturation_vapour_pressure(void);
	float slope_of_saturation_vapour_pressure(float);
	float actual_vapour_pressure(void);
	float vapour_pressure_deficit(void);
	float relative_sun_distance(void);
	float solar_declination(void);
	float X(void);
	float sunset_hour_angle(void);
	float R_a(void);
	float R_a_in_mm(void);
	float daylight_hours(void);
	float solar_radiation(void);
	float solar_radiation_in_mm(void);
	float R_so(void);
	float R_ns(void);
	float R_nl(void);
	float net_radiation(void);
	float net_radiation_in_mm(void);
	float RH(float);
	float soil_heat_flux(void);
	float eto_hargreaves(void);
	float eto(void);

};

#endif /* LIBPENMON_H_ */


libpenmon.cpp
//============================================================================
// Name        : libpenmon.cpp
// Author      : Sherzod RUZMETOV
// Version     :
// Copyright   : 2020 Sherzod RUZMETOV
// Description : Implements Weather Station that calculates ETo using
//               Penman-Monteith equation
//============================================================================

#include "libpenmon.h"
#include <math.h>
#include <assert.h>
#include <iostream>

using namespace std;

bool is_set(float n) {
	return (n != -999.99);
}

bool is_set(int n) {
	return (n != -999);
}

Crop::Crop(void) {
	resistance_a = 208;
	albedo = 0.23;
	height = 0.12;
}

Climate::Climate(void) {

	interior_location = true;
	coastal_location = false;
	island_location = false;

	arid_climate = true;
	humid_climate = false;

	dew_point_difference = 2;
	average_wind_speed = 2.0;
}

void Climate::light_winds() {
	this->average_wind_speed = 0.5;
}
void Climate::moderate_winds() {
	this->average_wind_speed = 2.0;
}
void Climate::strong_winds() {
	this->average_wind_speed = 4.0;
}

void Climate::arid() {
	this->arid_climate = true;
	this->humid_climate = false;
	this->dew_point_difference = 2;
}

void Climate::humid() {
	this->arid_climate = false;
	this->humid_climate = true;
	this->dew_point_difference = 0;
}

void Climate::interior() {
	this->interior_location = true;
	this->coastal_location = false;
	this->island_location = false;
	this->k_rs = 0.16;
}

void Climate::coastal() {
	this->interior_location = false;
	this->coastal_location = true;
	this->island_location = false;
	this->k_rs = 0.19;
}

void Climate::island() {
	this->interior_location = false;
	this->coastal_location = false;
	this->island_location = true;
	this->k_rs = 0.19;
}

Station::Station(float lat, int alt) {

	this->latitude = lat;
	this->altitude = alt;
	this->latitude_radians = ( M_PI / 180) * this->latitude;
	this->ref_crop = Crop();
	this->climate = Climate();
}

Station::Station(float lat, int alt, short anem_height) {
	this->latitude = lat;
	this->altitude = alt;
	this->anemometer_height = anem_height;
	this->latitude_radians = (M_PI / 180) * this->latitude;
	this->ref_crop = Crop();
	this->climate = Climate();
}

StationDay Station::get_day(short d) {

	StationDay day(d, this);

	return day;
}

StationDay::StationDay(short d, Station *st) {
	this->day_number = d;
	this->station = st;
}

float StationDay::wind_speed_2m() {

	if (this->wind_speed && (this->station->anemometer_height == 2)) {
		return this->wind_speed;
	}

	if (this->wind_speed && (this->station->anemometer_height != 2)) {
		float converted_wind_speed =
				this->wind_speed
						* (4.87
								/ log(
										(67.8 * this->station->anemometer_height)
												- 5.42));

		return converted_wind_speed;
	}

	if (this->station->climate) {
		return this->station->climate.average_wind_speed;
	}
	return -1.0;
}

float StationDay::dew_point() {
	if (this->temp_dew) {
		return this->temp_dew;
	}
	if (this->temp_min && this->station->climate.dew_point_difference) {
		return this->temp_min - this->station->climate.dew_point_difference;
	}
	return false;
}

float StationDay::atmospheric_pressure() {
	return 101.3 * pow((293 - (0.0065 * this->station->altitude)) / 293, 5.26);
}

float StationDay::atmospheric_pressure_in_mm() {
	return this->atmospheric_pressure() * 7.50;
}

float StationDay::latent_heat_of_vapourization() {
	return 2.45;
}

double StationDay::specific_heat() {
	return 1.013 * pow(10, -3);
}

float StationDay::psychrometric_constant() {
	return 0.665 * pow(10, -3) * this->atmospheric_pressure();
}

float StationDay::Tmean() {

	if (this->temp_mean) {
		return this->temp_mean;
	}
	if (this->temp_max && this->temp_min) {
		return (this->temp_max + this->temp_min) / 2;
	}
	return false;
}

float StationDay::saturation_vapour_pressure(float T) {
	return 0.6108 * pow(2.7183, ((17.27 * T) / (T + 237.3)));
}

float StationDay::mean_saturation_vapour_pressure() {
	if (this->temp_max && this->temp_min) {
		float vp_max = this->saturation_vapour_pressure(this->temp_max);
		float vp_min = this->saturation_vapour_pressure(this->temp_min);
		return (vp_max + vp_min) / 2;
	}

	if (this->temp_mean) {
		return this->saturation_vapour_pressure(this->temp_mean);
	}
	return false;
}

float StationDay::slope_of_saturation_vapour_pressure(float T) {
	return 4098 * 0.6108 * pow(2.7183, (17.27 * T) / (T + 237.3))
			/ pow((T + 237.3), 2);
}

float StationDay::actual_vapour_pressure() {
	if (this->vapour_pressure) {
		return this->vapour_pressure;
	}
	if (this->temp_dry && this->temp_wet) {
		float vp_wet = this->saturation_vapour_pressure(this->temp_wet);
		float psychrometric_constant = this->psychrometric_constant();
		return vp_wet
				- (psychrometric_constant * (this->temp_dry - this->temp_wet));
	}
	if (this->humidity_max && this->humidity_min && this->temp_max
			&& this->temp_min) {
		float vp_min = this->saturation_vapour_pressure(this->temp_min);
		float vp_max = this->saturation_vapour_pressure(this->temp_max);
		return (vp_min * (this->humidity_max / 100)
				+ vp_max * (this->humidity_min / 100)) / 2;
	}
	if (this->humidity_max && this->temp_min) {
		float vp_min = this->saturation_vapour_pressure(this->temp_min);
		return vp_min * (this->humidity_max / 100);
	}
	if (this->humidity_mean && this->temp_max && this->temp_min) {
		float vp_min = this->saturation_vapour_pressure(this->temp_min);
		float vp_max = this->saturation_vapour_pressure(this->temp_max);
		return (this->humidity_mean / 100) * ((vp_max + vp_min) / 2);
	}
	if (this->dew_point()) {
		return this->saturation_vapour_pressure(this->dew_point());
	}
	return false;
}

float StationDay::vapour_pressure_deficit() {

	if (this->temp_min && this->temp_max) {
		float vp_min = this->saturation_vapour_pressure(this->temp_min);
		float vp_max = this->saturation_vapour_pressure(this->temp_max);
		float vp_actual = this->actual_vapour_pressure();
		return ((vp_min + vp_max) / 2) - vp_actual;
	}
	return false;
}

float StationDay::relative_sun_distance() {
	return (1 + (0.033 * cos((2 * M_PI / 365) * this->day_number)));
}

float StationDay::solar_declination() {
	return 0.409 * sin(((2 * M_PI / 365) * this->day_number) - 1.39);
}

float StationDay::X() {
	float x = 1
			- (tan(this->station->latitude_radians)
					* tan(this->solar_declination()));
	if (x <= 0) {
		x = 0.00001;
	}
	return x;
}

float StationDay::sunset_hour_angle() {
	return acos(
			-1 * tan(this->station->latitude_radians)
					* tan(this->solar_declination()));
}

float StationDay::R_a() {

	float lat = this->station->latitude_radians;
	float decl = this->solar_declination();
	float angle = this->sunset_hour_angle();
	double P = (angle * sin(lat) * sin(decl))
			+ (cos(lat) * cos(decl) * sin(angle));

	return (24 * 60 / M_PI) * 0.0820 * this->relative_sun_distance() * P;
}

float StationDay::R_a_in_mm() {
	return this->R_a() * 0.408;
}

float StationDay::daylight_hours() {
	return (24 / M_PI) * this->sunset_hour_angle();
}

float StationDay::R_so() {
	int alt = this->station->altitude;
	if (alt < 100) {
		return (0.25 + 0.50) * this->R_a();
	}
	return (0.75 + (pow(210, -5) * alt)) * this->R_a();
}

float StationDay::solar_radiation() {
	if (this->radiation_s) {
		if (CHECK_RADIATION_RANGE && (this->radiation_s > this->R_so())) {
			throw("RADIATION_RANGE");
		}
		return this->radiation_s;
	}

	// IMPORTANT: sunshine_hours will be negative if it's not initialized
	float n = this->sunshine_hours;

	if (n < 0) {
		// it means sunshine_hours is NOT set
		if (this->station->climate && this->station->climate.island_location) {
			return (0.7 * this->R_a()) - 4;
		}

		if (this->station->climate and this->temp_min && this->temp_max) {
			float krs = 0.16;
			if (this->station->climate.coastal_location) {
				krs = 0.16;
			}
			return krs * sqrt(this->temp_max - this->temp_min) * this->R_a();
		}
		n = this->daylight_hours();
	} // n < 0

	if (CHECK_SUNSHINE_HOURS_RANGE && (n > this->daylight_hours())) {
		throw("SUNSHINE_RANGE");
	}

	return (0.25 + (0.50 * (n / this->daylight_hours()))) * this->R_a();
}

float StationDay::solar_radiation_in_mm() {
	return (this->solar_radiation() * 0.408);
}

float StationDay::R_ns() {
	return (1 - this->station->ref_crop.albedo) * this->solar_radiation();
}

float StationDay::R_nl() {

	float TmaxK = this->temp_max + 273.16;
	float TminK = this->temp_min + 273.16;
	float ea = this->actual_vapour_pressure();
	float rs = this->solar_radiation();
	float rso = this->R_so();
	float sbc = this->stephen_boltzmann_constant;

	return sbc * ((pow(TmaxK, 4) + pow(TminK, 4)) / 2) * (0.34 - (0.14 * sqrt(ea)))
			* ((1.35 * (rs / rso)) - 0.35);
}

float StationDay::net_radiation() {
	if (this->R_ns() && this->R_nl()) {
		return this->R_ns() - this->R_nl();
	}
	return false;
}

float StationDay::net_radiation_in_mm() {
	if (this->net_radiation()) {
		return this->net_radiation() * 0.408;
	}
	return false;
}

float StationDay::RH(float T) {
	return 100
			* (this->actual_vapour_pressure()
					/ this->saturation_vapour_pressure(T));
}

float StationDay::soil_heat_flux() {
	return 0.00;
}

float StationDay::eto_hargreaves() {
	assert(this->temp_max && this->temp_min);

	float Tmean = (this->temp_max + this->temp_min) / 2;
	return (Tmean + 17.8) * pow((this->temp_max - this->temp_min), 0.5)
			* this->R_a();
}

float StationDay::eto() {

	if (this->wind_speed_2m() < 0) {
		return this->eto_hargreaves();
	}

	float Tmax = this->temp_max;
	float Tmin = this->temp_min;
	float Tmean = this->temp_mean;

	if (Tmax && Tmin) {
		Tmean = (Tmax + Tmin) / 2;
	}

	float slope = this->slope_of_saturation_vapour_pressure(Tmean);
	float net_radiation = this->net_radiation();
	float G = this->soil_heat_flux();
	float u2m = this->wind_speed_2m();

	//cout << "Net radiation is " << net_radiation << endl;

	float nominator = (0.408 * slope * (net_radiation - G))
			+ (this->psychrometric_constant() * (900 / (Tmean + 273)) * u2m
					* this->vapour_pressure_deficit());
	float denominator = slope
			+ (this->psychrometric_constant() * (1 + (0.34 * u2m)));

	return nominator / denominator;
}



example.cpp
/* * main.cpp
 *
 *  	Created on: Oct 9, 2020
 *      Author: sherzodr
 */

#include <iostream>
#include "libpenmon.h"
#include <string>
#include <stdio.h>
#include <string.h>

using namespace std;

void testEq(float a, float b, string message) {

	char a_str[50];
	char b_str[50];

	sprintf(a_str, "%.3f", a);
	sprintf(b_str, "%.3f", b);

	if (strcmp(a_str, b_str) == 0) {
		cout << message << "... \e[1;92mPASSED\e[0m";
		//cout << message << "... PASSED";
	} else {
		cout << message << "... \e[1;31mNOT PASSED. Expected " << b_str << ". Received  " << a_str << "\e[0m";
		//cout << message << "... NOT PASSED. Expected " << b_str
		//		<< ". Received  " << a_str;
	}
	cout << "\n";
}

void testTrue(bool param, string message) {
	if (param == true) {
		cout << message << "... \e[1;92mPASSED\e[0m";
	} else {
		cout << message << "... \e[1;31mNOT PASSED\e[0m";
	}
	cout << endl;
}

int main(void) {

	Station station(41.42, 109);
	StationDay day = station.get_day(130);

	testEq(station.latitude, 41.42, "testing for latitude");
	testEq(station.altitude, 109.00, "testing for altitude");
	testEq(station.anemometer_height, 2.0, "anemometer height");
	testEq(day.day_number, 130, "day_number");

	Station station_1080(41.42, 1800);
	StationDay day_130 = station_1080.get_day(130);
	testEq(day_130.atmospheric_pressure(), 81.756, "atmospheric pressure");
	testEq(day_130.psychrometric_constant(), 0.054397,
			"psychrometric_constant");
	testEq(day.specific_heat(), pow(1.013 * 10, -3), "specific heat");
	testEq(day.latent_heat_of_vapourization(), 2.45,
			"latent_heat_of_vapourziation");

	day.temp_max = 24.5;
	day.temp_min = 15;
	testEq(day.mean_saturation_vapour_pressure(), 2.39,
			"mean saturation vapour pressure");

	day.temp_dew = 19.5;
	testEq(day.RH(35), 40.316, "RH at 35 degrees");

//	Station station2(41.42, 109);
//	StationDay day2=station2.get_day(130);
//	day2.temp_mean=19.75;

	day.temp_max = false;
	day.temp_min = false;
	day.temp_dew = false;
	day.temp_mean = 19.75;
	testEq(day.mean_saturation_vapour_pressure(), 2.302,
			"mean saturation vapour pressure ");

	day.temp_mean = false;
	testEq(day.slope_of_saturation_vapour_pressure(24.5), 0.183837,
			"slove of vapour pressure curve");

	// actual pscychrometric vapour pressure

	station.altitude = 1200;
	day.temp_wet = 19.5;
	day.temp_dry = 25.6;
	testEq(day.atmospheric_pressure(), 87.897, "atmospheric pressure");
	testEq(day.saturation_vapour_pressure(19.5), 2.267,
			"saturation vapour pressure");
	testTrue(day.temp_dew != true, "Temp dew temperature is NOT set");
	testEq(day.actual_vapour_pressure(), 1.91, "Actual vapour pressure");

	day.temp_dew = 17.0;
	day.temp_wet = false;
	day.temp_dry = false;

	testEq(day.temp_dew, 17.0, "Dew temp is set");
	testEq(day.actual_vapour_pressure(), 1.938, "Actual vapour pressure");

	day.temp_dew = false;
	day.temp_min = 18;
	day.temp_max = 25;
	day.humidity_max = 82;
	day.humidity_min = 54;

	testEq(day.saturation_vapour_pressure(18), 2.064,
			"saturation vapour pressure");
	testEq(day.saturation_vapour_pressure(25), 3.168,
			"saturation vapour pressure");
	testEq(day.actual_vapour_pressure(), 1.702, "actual vapour pressure");

	day.temp_min = 18;
	day.humidity_max = 82;
	day.temp_dew = false;
	day.temp_max = false;
	day.humidity_min = false;

	testEq(day.actual_vapour_pressure(), 1.692, "actual vapour pressure");

	day.humidity_min = false;
	day.temp_dew = false;
	day.humidity_max = false;
	day.temp_min = 18;
	day.temp_max = 25;
	day.humidity_mean = 68;

	testEq(day.actual_vapour_pressure(), 1.779,
			"actual vapour pressure is 1.779");

	day.humidity_mean = false;
	day.temp_min = 18;
	day.temp_max = 25;
	day.humidity_max = 82;
	day.humidity_min = 54;

	testEq(day.vapour_pressure_deficit(), 0.914,
			"determining vapour pressure deficit");

	Station station_bangkok(13.73, 1200);
	testEq(station_bangkok.latitude_radians, 0.240, "latitude radians");

	Station station_rio(22.90, 1200);
	testEq(station_rio.latitude_radians, 0.400, "latitude radians");

	StationDay day_246 = station.get_day(246);
	testEq(day_246.relative_sun_distance(), 0.985, "relative sun distance");
	testEq(day_246.solar_declination(), 0.120, "solar declination");

	Station south_station(-20.0, 1200);
	StationDay south_day = south_station.get_day(246);

	testEq(south_day.sunset_hour_angle(), 1.527, "sunset hour angle");
	testEq(south_day.R_a(), 32.194, "R_a");
	testEq(south_day.R_a_in_mm(), 13.135, "R_a_in_mm");
	testEq(south_day.daylight_hours(), 11.666, "daylight hours");

	Station south_station2(-22.90, 1200);
	StationDay day135 = south_station2.get_day(135);
	day135.sunshine_hours = 7.10;
	testEq(day135.solar_radiation(), 14.46, "solar radiation");
	testEq(day135.R_so(), 18.833, "clear sky radiation");

	testEq(day135.R_ns(), 11.134, "net shortwave radiation");

	day135.temp_min=25.1;
	day135.temp_max=19.1;
	day135.vapour_pressure=2.1;
	testEq(day135.actual_vapour_pressure(), 2.1, "actual vapour pressure");
	day135.sunshine_hours=7.1;
	testEq(day135.R_nl(), 3.51, "Net longwave raditaion");
	testEq(day135.net_radiation(), 7.624, "net radiation");

	day135.wind_speed=5;
	testEq(day135.wind_speed_2m(), 5, "wind speed" );

	day135.station->anemometer_height=10;
	day135.wind_speed=3.2;
	testEq(day135.wind_speed_2m(), 2.393, "wind_speed");

	Station qarmish(41.42, 109);
	StationDay day150 = qarmish.get_day(150);
	day150.temp_min = 19.5;
	day150.temp_max = 36.50;
	day150.wind_speed = 2;
	testEq(day150.eto(), 6.98, "eto()");

	return 0;
}

// test
